src/components/zoneddatetime.rs

```diff
@@ -400,6 +420,15 @@ impl ZonedDateTime {
 
 #[cfg(feature = "experimental")]
 impl ZonedDateTime {
+    /// Creates a new `ZonedDateTime` from the current `ZonedDateTime`
+    /// combined with the provided `TimeZone`.
+    pub fn with_plain_time(&self, time: Option<PlainTime>) -> TemporalResult<Self> {
```

Comment:
Hmm, does it make sense to take `Option` here? I would assume if someone's calling `with_plain_time`, they would have a specific time to set the datetime to

Reply:
Yeah, it's a good note.

The [proposal](https://tc39.es/proposal-temporal/#sec-temporal.zoneddatetime.prototype.withplaintime) is fairly clear that the argument for `with_plain_time` is optional and can be undefined. So from a proposal perspective, this is technically correct. But to what degree does it makes sense as an API ...

Hmmmm, sort of thinking out loud. `with_plain_time` is on some level `with_epoch_nanoseconds` where the `IsoDate` of the `ZonedDateTime` is a base.

We could have a `with_epoch_nanosecond` and then reconstruct the functionality in Boa using that method.

From Boa's side, it would probably end up looking like:

```
let zdt = this.downcast_ref ...
let time = args.first().map(|v| to_temporal_time(v)).transpose()?;

let epoch_nanos = if let Some(pt) = time {
    // `as_epoch_nanoseconds` would need to be added to `PlainDateTime`.
    zdt.to_plain_datetime().with_time(pt)?.as_epoch_nanoseconds()
} else {
   zdt.start_of_day(context.tz_provider())?
};

ZonedDateTimeInner::with_epoch_nanoseconds(epoch_nanos)
```

I don't necessarily dislike it ... but it also brings up the question to a point of if it's worthwhile having a `with_epoch_nanoseconds` and `with_instant`.

Thoughts?

Reply:
Or alternatively, we could just support the path where `PlainTime` is available.

```rust
fn with_plain_time(this: &JsValue, args: [JsValue], context: &mut Context) -> JsResult<JsValue> {
    let zdt = this.
        .as_object()
        .and_then(JsObject::downcast_ref::<Self>)
        .ok_or_else( // -- throw error -- //)?;

    let time = args.first().map(|v| to_temporal_time(v)).transpose()?;
    
    if let Some(pt) = time {
        zdt.inner.with_time(pt, context.tz_provider())
    } else {
       zdt.inner.with_instant(zdt.start_of_day(context.tz_provider())?.into())
    }
}
```

Reply:
Yeah, the second one looks a bit cleaner